cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+                                        uint64_t vtid, char* name) {
+  std::string name_str(name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  state->entry_kernel_name[hpt] = name_str;
+}
+
+void module_get_function_exit_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult,
+                                       CUfunction cuFunction) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  try {
+    auto kernel_name = state->entry_kernel_name.at(hpt);
+    hp_kernel_t hp_kernel_key = {hostname, vpid, cuFunction};
+    state->hp_kernel_to_name[hp_kernel_key] = kernel_name;
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "entry kernel name not found in module get function exit callback");
+  }
+}
+
+void operation_entry_helper(void *btx_handle, void *usr_data,
+                            int64_t ts, const char *event_class_name,
+                            const char *hostname, int64_t vpid,
+                            uint64_t vtid, CUstream cuStream, CUfunction f) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  CUdevice dev;
+  bool found = state->context_manager.get_stream_device(hpt, cuStream, &dev);
+  if (!found) {
+    THAPI_FATAL_HPT(hpt, "no device for thread in device op");
+  }
+
+  std::string name = strip_event_class_name(event_class_name);
+  if (f == nullptr) {
+    // for host launch and non launch operations like cuMemcpy, the device op name
+    // is just the function name
+    state->hpt_function_name_to_dev[{hostname, vpid, vtid, name}] = dev;
+    state->hpt_profiled_function_name_and_ts[hpt] = fn_ts_t(name, ts);
+  } else {
+    // for device launch, get the name saved when the kernel was loaded in the
+    // module get function callbacks
+    try
+    {
+      std::string name = state->hp_kernel_to_name.at({hostname, vpid, f});
+      state->hpt_function_name_to_dev[{hostname, vpid, vtid, name}] = dev;
+      state->hpt_profiled_function_name_and_ts[hpt] = fn_ts_t(name, ts);
+    } catch(const std::out_of_range& oor) {
+      THAPI_FATAL_HPT(hpt,
+        "device not found for hpt function name in kernel launch callback");
+    }
+  }
+}
+
+void named_operation_stream_present_entry_callback(void *btx_handle, void *usr_data,
+                                                   int64_t ts, const char *event_class_name,
+                                                   const char *hostname, int64_t vpid,
+                                                   uint64_t vtid, CUstream cuStream) {
+  operation_entry_helper(btx_handle, usr_data, ts, event_class_name,
```

Comment:
Not a good name, not really explicit

Reply:
I changed to use `task` and added some comments, let me know if new code is clear enough.

Reply:
Also use "kernel task" and "non kernel task" terminology to distinguish the two, I think is a lot clearer. The kernel ones have the CUfunction arg, everything else does not and uses the task launch function name for the emitted event name.

Reply:
Nice job!